File: Datamodeler.m
(* ::Package:: *)

listtable = {}
parameters = {}
If[$CommandLine[[5]] == "-f", Do[listtable = Append[listtable, Internal`StringToDouble@ToString[$CommandLine[[i + 5]]]],{i,Length[$CommandLine[[]]] - 5}]]
(* If there are parameters, Then... *)
If[$CommandLine[[5]] == "-p", 
(* Then TRUE *)
(* Grabs -f data.  If -f is present (or if -f is not empty in set) *)
If[Position[$CommandLine, "-f"]!= {},
(* Parameters *)
Do[parameters = Append[parameters, Internal`StringToDouble@ToString[$CommandLine[[i + 5]]]],
{i,Round[Internal`StringToDouble@ToString[Position[$CommandLine, "-f"]]] - 6}]
(* Data *)
Do[listtable = Append[listtable, Internal`StringToDouble@ToString[$CommandLine[[i + Round[Internal`StringToDouble@ToString[Position[$CommandLine, "-f"]]]]]]],
{i,Length[$CommandLine[[]]] - Round[Internal`StringToDouble@ToString[Position[$CommandLine, "-f"]]]}]]

(* Grabs -d data. If -d is present (or if -d is not empty in set) *)
If[Position[$CommandLine, "-d"]!= {},
(* Parameters *)
Do[parameters = Append[parameters, Internal`StringToDouble@ToString[$CommandLine[[i + 5]]]],
{i,Round[Internal`StringToDouble@ToString[Position[$CommandLine, "-d"]]] - 6}]
(* Data *)
Do[listtable = Append[listtable, Internal`StringToDouble@ToString[$CommandLine[[i + Round[Internal`StringToDouble@ToString[Position[$CommandLine, "-d"]]]]]]],
{i,Length[$CommandLine[[]]] - Round[Internal`StringToDouble@ToString[Position[$CommandLine, "-d"]]]}]]

,
(* Then FALSE *)
Do[listtable = Append[listtable, Internal`StringToDouble@ToString[$CommandLine[[i + 5]]]],{i,Length[$CommandLine[[]]] - 5}]
]
Print[parameters]
Print[listtable]

(* DATAMODELER CODE GOES HERE BASED OFF OF LISTTABLE AND PARAMETERS *)
If[parameters == {}, 
(* Parameters are not used and symbolic regression is used with basic data input *)

,
(* Parameters used, if numeric, used for restraints, otherwise it could be to run
symbolic regression a certain way *)

]

File: Datamodeler.py
#Program calls Datamodeler.m with given arguements.
#Python Datamodeler.py <table of elements> <Time?> find out more.
#Author: Kyler Schmidt
from subprocess import *
from sys import *

#INITALIZATION
#/***********************************************************************/
#File called with mathematica script arguments.
#Assuming same directory as Datamodeler.py (required to run)
command='-script Datamodeler.m'

#All arguments stored into parameter. First arguement is ignored
#Since it is the file name
parameter=str(argv[1:])
#/***********************************************************************/

#GRAB FLAGS USED HERE
flags_used = " "
if "-f" in parameter:
    flags_used = "-f"
if "-d" in parameter:
    flags_used = flags_used + " -d"
if "-p" in parameter:
    flags_used = flags_used + " -p"
if "-h" in parameter:
    flags_used = flags_used + " -h"
if "-f -d" in flags_used:
    print(flags_used)
    print ("file and data cannot be used at the same time.")
    exit()
if "-f" not in flags_used and "-d" not in flags_used and "-h" not in flags_used:
    print("No data input detected. refer to -h for input")
    exit()

    
#FLAGS
#/***********************************************************************/
#help flag - stops code.
if "-h" in parameter:
#   Datamodeler.py useage
    print ("\n\nusage: Datamodeler.py [-h] [-p] [-f] [-d] ",
           "\nDatamodeler.m must be in the same directory as Datamodeler.py",
           "\nAlways use flag -p before flags -d or -f. other ways may not be supported.",
           "\n\n-h: Displays this help guide",
           "\n-f: input file for data. input parameters separately or in file provide an -p with numbers.",
           "\n-d: direct data input into command line",
           "\n\nThe following are examples on how to use each flag:",
           "\n\n: -h: Datamodeler.py -h",
           "\n: -f: Datamodeler.py -f Data.txt",
           "\n: -p and -d: Datamodeler.py -p 1 2 3 4 -d 1 2 3 4",)
    exit()

#Check for if arguments are present past the -h flag
if len(argv) < 3:
    print ("usage: Datamodeler.py [-h] [-f] [-d] [-p] arguments")
    exit()

#if file flag is used
if "-f" in parameter:
#   Better flag integerations. Splits up string after certain flag.
#   Use this for more flag integerations with
    try:
        file_name = parameter.split("'-f', '",1)[1]
        previous_data = parameter.split("'-f',",1) [0]
    except IndexError:
        print ("Error Occured with flag -f. Is it separated with spaces?")
        exit()
    file_name = file_name.split("']",1) [0]
    
    if "." not in file_name:
        file_name = file_name + ".txt"
        
    try:
        fp = open(file_name)
    except FileNotFoundError:
        print ("File not found in directory: " + file_name)
        exit()
    
    parameter = previous_data + "-f " + fp.read()
#/***********************************************************************/


#REFORMAT
#/***********************************************************************/
parameter=parameter.replace('[','')
parameter=parameter.replace(',','')
parameter=parameter.replace("'",'')
parameter=parameter.replace(']','')
#/***********************************************************************/

#Used to correct if only calling with -f. Breaks for some reason...
if "-p" not in parameter and "-f" in parameter:
    parameter = "-p " + parameter

#grab argument data and send over to Datamodeler.m.
call(['wolframscript',command, parameter])
